### ISSUE TITLE
radical: exception chaining

### ISSUE THREAD

**Issue by andre-merzky** [radical-cybertools/radical.utils#23]
Python allows this:

```
try :
  1/0
except Exception as e :
  raise my.NoSuccess ("oops") from e
```

which preserves the original exception stack.  We need to do this for all exceptions.  Also, this is recommended in PEP8, for those who care.

**Comment by andre-merzky** [comment]
```
pip install future
```

```
from future.utils import raise_from

class Test:
    def __init__(self):
        try:
            1/0
        except Exception as e:
            raise_from(RuntimeError('failed'), e)
```

**Comment by andre-merzky** [comment]
This is now addressed in #210 - but should also be applied to RS, RP and RE.

**Comment by mturilli** [comment]
Wow, 6 years! Well, "who persists, wins!" :D Yes, let's do the same for the whole stack. A good opportunity also to have a look at exceptions in general?

**Comment by andre-merzky** [comment]
> A good opportunity also to have a look at exceptions in general?

Is there need to do so?  The SAGA layer has custom exceptions (per spec), but is fairly consistent about them.  RP and RU got rid of custom exceptions: the lack of proper definition and inconsistent use made them rather useless.  EnTK still has custom exceptions - I don't think they are needed, but don't feel strongly about it at all.  Did you have any other points in mind?

**Comment by mturilli** [comment]
I am thinking about customized and more readable error/warning messages to the user. The example that comes to mind is when a mongodb has not been configured. After many years, I recognize the trace but (as it has happened a couple of times lately) users have no idea of what is happening.

**Comment by andre-merzky** [comment]
Ah, right, that is a perfectly valid concern!   That's out of scope for this round of PRs (exception chaining), but needs to be addressed.  For the record, I think that should be addressed like this:

  - all components (and indirectly agent) catch all exceptions on the highest layer (component main loop, thread main loop, etc.).  That already happens for logging purposes (in most cases)
  - caught exceptions are translated into ZMQ messages and sent on the `LOG` ZMQ channel
  - the session instance within the application subscribes to the `LOG` channel - this way the error messages arrive at the application.
  - the application can either install log callbacks for critical errors (`session.set_error_cb(my_cb)`), or we can configure the session to terminate on critical errors (`session = rp.Session(..., close_on_error=True)`)

That way all (relevant) errors should arrive in the application process, instead of getting buried in the log files.  There are two types of errors for which this won't work: ZMQ messaging errors and bootstrapping errors (those happen before we can register to the `LOG` channel).  Oh well...

That doesn't make messages readable though.  But have a look at [this](https://github.com/radical-cybertools/radical.saga/blob/devel/src/radical/saga/utils/pty_exceptions.py#L12) in the SAGA pty layer.  I envision something similar for RP: out of the infinite possible set of errors / exceptions, derive possible causes and remedies heuristically from the origin of the error and from message inspection.  If a component can do that translation before sending the error on the log channel, so the better (interpretation stays more local to the cause), but that might not always be possible,

We have most pieces in place for this (components catch errors, the `LOG` channel exists) - we miss ZMQ between agent and client to make this seamless, and miss the respective exception translators (or, more precisely, the time and motivation to implement them, as that is rather tedious business).

That answer is probably longer than you expected - but let me know what you think about that approach :-)

**Comment by mturilli** [comment]
Sounds somehow _ad hoc_ but I like it :) I think it smartly addresses the problem of distributing exceptions in absence of a native/system way to do that. More conventional would be perhaps to assume a centralized DB instead of distributed communication/coordination? If so, I think that would be a general design decision that we agreed to evaluate in the design phase of RP2. Should we C&P your message into a dedicated ticket, maybe marked as RP2? Note also that this requires an answer to: are we going to extend RP1 or starts anew with RP2? I think extending here would be useful to test the idea and see whether it works well?
